### ISSUE TITLE
UnsatisfiedLinkError, libc.so.6 error and that stuff...

### ISSUE THREAD

**Issue by ElRafeta697** [xerial/sqlite-jdbc#659]
I tried to make a simple app to query SQLite databases with SQLite JDBC 3.23.1, but this doesn't care because it happens with every version. The code is so easy: I get a connection with `DriverManager.getConnection` and i refer to my `jdbc:sqlite:database.db` file. But the problem comes when this simple sentence drops the next error:
```
java.lang.UnsatisfiedLinkError: dlopen failed: library "libc.so.6" not found
        at java.lang.Runtime.loadLibrary0(Runtime.java:1071)
        at java.lang.Runtime.loadLibrary0(Runtime.java:1007)
        at java.lang.System.loadLibrary(System.java:1667)
        at org.sqlite.core.NativeDB.<clinit>(NativeDB.java:38)
        at org.sqlite.core.NativeDB.load(NativeDB.java:53)
        at org.sqlite.core.CoreConnection.open(CoreConnection.java:211)
        at org.sqlite.core.CoreConnection.<init>(CoreConnection.java:76)
        at org.sqlite.jdbc3.JDBC3Connection.<init>(JDBC3Connection.java:25)
        at org.sqlite.jdbc4.JDBC4Connection.<init>(JDBC4Connection.java:24)
        at org.sqlite.SQLiteConnection.<init>(SQLiteConnection.java:45)
        at org.sqlite.JDBC.createConnection(JDBC.java:114)
        at org.sqlite.JDBC.connect(JDBC.java:88)
        at java.sql.DriverManager.getConnection(DriverManager.java:580)
        at java.sql.DriverManager.getConnection(DriverManager.java:236)
        at edu.androidstudio.sqlide.Classes.SQLConnection.connect(SQLConnection.java:32)
        at edu.androidstudio.sqlide.Views.FileExplorer.onClick(FileExplorer.java:102)
        at edu.androidstudio.sqlide.Classes.FileExplorerController$FileExplorerItem.onClick(FileExplorerController.java:115)
        at android.view.View.performClick(View.java:7259)
        at android.view.View.performClickInternal(View.java:7236)
        at android.view.View.access$3600(View.java:801)
        at android.view.View$PerformClick.run(View.java:27892)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
```
I found than many people had the same problem, and a guy (@gaorongfu) mentions that's because this driver not supports 64bit android systems, but this issue is from July of 2020 and maybe can be buried by another issues.

So I asked this with two purposes:
 - To search information about the problem and learn (I'm not sure if the problem comes from the architecture like the mentioned guy (@gaorongfu) said)
 - If @gaorongfu is right, to dig up this issue, because I think it's important to solve due to the fact that fewer and fewer devices have 32 bit architecture and this is a good alterntive to native Android SQLite driver.

**Comment by gotson** [comment]
@xerial I see dockcross now has android images for arm, arm64, x86 and x86_64. Should we update the logic in OSInfo to prepend `android-` to the detected arch, but use the normal architecture detection? That way we could support 4 architectures for Android.

**Comment by ElRafeta697** [comment]
Maype be worth it. Depending of the OS, change the comprobation

**Comment by gotson** [comment]
> In addition, Despite SQLite can be used in many platforms, it's main target is usually Android

That's just your assumption, probably totally wrong 🙄

**Comment by gotson** [comment]
@ElRafeta697 what's the android device you had the problem on? What cpu architecture is it? Armv8 (64 bits) is compatible with armv7 (32 bits) binaries. Is that an x86_64 cpu?

**Comment by ElRafeta697** [comment]
I used the jar of this project (I think I have the 3.23.1 version), and i get the libsqlitejdbc.so from inside the jar. With a x86 emulator (genymotion) i get the libc.so.6 error, and with an arm64-v8a the app cannot found libsqlitejdbc.so

**Comment by gotson** [comment]
> I used the jar of this project (I think I have the 3.23.1 version), and i get the libsqlitejdbc.so from inside the jar. With a x86 emulator (genymotion) i get the libc.so.6 error, and with an arm64-v8a the app cannot found libsqlitejdbc.so

For x86 it's expected but there's only a couple phones with x86 processors. Can you post the error with Armv8?

**Comment by ElRafeta697** [comment]
And you think Genymotion_ARM_Translation could help me?


And the error I get with armv8 is 
```
java.lang.UnsatisfiedLinkError: dlopen failed: library "libsqlitejdbc.so" not found
        at java.lang.Runtime.loadLibrary0(Runtime.java:1087)
        at java.lang.Runtime.loadLibrary0(Runtime.java:1008)
        at java.lang.System.loadLibrary(System.java:1664)
        at org.sqlite.core.NativeDB.<clinit>(NativeDB.java:39)
        at org.sqlite.core.NativeDB.load(NativeDB.java:60)
        at org.sqlite.SQLiteConnection.open(SQLiteConnection.java:240)
        at org.sqlite.SQLiteConnection.<init>(SQLiteConnection.java:61)
        at org.sqlite.jdbc3.JDBC3Connection.<init>(JDBC3Connection.java:28)
        at org.sqlite.jdbc4.JDBC4Connection.<init>(JDBC4Connection.java:21)
        at org.sqlite.JDBC.createConnection(JDBC.java:115)
        at org.sqlite.JDBC.connect(JDBC.java:90)
        at java.sql.DriverManager.getConnection(DriverManager.java:580)
        at java.sql.DriverManager.getConnection(DriverManager.java:236)
        at edu.androidstudio.sqlide.Classes.SQLConnection.connect(SQLConnection.java:32)
        at edu.androidstudio.sqlide.Views.FileExplorer.onClick(FileExplorer.java:102)
        at edu.androidstudio.sqlide.Classes.FileExplorerController$FileExplorerItem.onClick(FileExplorerController.java:115)
        at android.view.View.performClick(View.java:8160)
        at android.view.View.performClickInternal(View.java:8137)
        at android.view.View.access$3700(View.java:888)
        at android.view.View$PerformClick.run(View.java:30236)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:246)
        at android.app.ActivityThread.main(ActivityThread.java:8595)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
```

**Comment by gotson** [comment]
> And you think Genymotion_ARM_Translation could help me?
> 
> And the error I get with armv8 is
> 
> ```
> java.lang.UnsatisfiedLinkError: dlopen failed: library "libsqlitejdbc.so" not found
>         at java.lang.Runtime.loadLibrary0(Runtime.java:1087)
>         at java.lang.Runtime.loadLibrary0(Runtime.java:1008)
>         at java.lang.System.loadLibrary(System.java:1664)
>         at org.sqlite.core.NativeDB.<clinit>(NativeDB.java:39)
>         at org.sqlite.core.NativeDB.load(NativeDB.java:60)
>         at org.sqlite.SQLiteConnection.open(SQLiteConnection.java:240)
>         at org.sqlite.SQLiteConnection.<init>(SQLiteConnection.java:61)
>         at org.sqlite.jdbc3.JDBC3Connection.<init>(JDBC3Connection.java:28)
>         at org.sqlite.jdbc4.JDBC4Connection.<init>(JDBC4Connection.java:21)
>         at org.sqlite.JDBC.createConnection(JDBC.java:115)
>         at org.sqlite.JDBC.connect(JDBC.java:90)
>         at java.sql.DriverManager.getConnection(DriverManager.java:580)
>         at java.sql.DriverManager.getConnection(DriverManager.java:236)
>         at edu.androidstudio.sqlide.Classes.SQLConnection.connect(SQLConnection.java:32)
>         at edu.androidstudio.sqlide.Views.FileExplorer.onClick(FileExplorer.java:102)
>         at edu.androidstudio.sqlide.Classes.FileExplorerController$FileExplorerItem.onClick(FileExplorerController.java:115)
>         at android.view.View.performClick(View.java:8160)
>         at android.view.View.performClickInternal(View.java:8137)
>         at android.view.View.access$3700(View.java:888)
>         at android.view.View$PerformClick.run(View.java:30236)
>         at android.os.Handler.handleCallback(Handler.java:938)
>         at android.os.Handler.dispatchMessage(Handler.java:99)
>         at android.os.Looper.loop(Looper.java:246)
>         at android.app.ActivityThread.main(ActivityThread.java:8595)
>         at java.lang.reflect.Method.invoke(Native Method)
>         at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
>         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
> ```

Do you have logs before that error? I think it should output which file it's trying to load

**Comment by ElRafeta697** [comment]
There aren't any logcat related to the database access, just throws the exception. But when I tried to go over the process with the debugger, Iread something like the classLoader was searching a file named classes6.dex or something like that. But I'm not sure

**Comment by xerial** [comment]
https://github.com/xerial/sqlite-jdbc/issues/659#issuecomment-913560418

Yeah. It seems OSInfo logic needs some updates to support android

**Comment by gotson** [comment]
@ElRafeta697 would you be able to test the fix in the PR #662 ? You would need to get the PR, build locally and install in MavenLocal, then use that version in your Android build.

**Comment by ElRafeta697** [comment]
Yes, I can, but I have to say I'm newbie to Github 😅

**Comment by gotson** [comment]
This is available in the latest snapshot, please give it a try.

**Comment by gotson** [comment]
Released as part of `3.39.2.0`
